In [ ]:
import pandas as pd
import importlib
import sys
import os

sys.path.append(
    os.path.abspath("../data_prepration")
)
import load_data
importlib.reload(load_data)

22
(16, 9)
#2B59C3


<module 'load_data' from 'e:\\DataAnalysis\\_Projects\\proj_17\\product_feature_performance_analysis\\data_prepration\\load_data.py'>

### **users table duplicates check**

In [7]:
users = load_data.load_users()
dups_count = users.duplicated().sum()
print (f"Duplicated Values Count: {dups_count}")

Duplicated Values Count: 0


#### Findings

No exact duplicate records were found in the users table.

### **users table missing value check**

In [16]:
missing_values_count = users.isna().sum().to_frame("Missing Values")
# print (f"Missing Values Count: {missing_values_count}")
missing_values_count["Total Rows"] = len(users)
missing_values_count["Percentage"] = (missing_values_count["Missing Values"] * 100 / missing_values_count["Total Rows"]).round(2)
missing_values_count

,Missing Values,Total Rows,Percentage
user_id,0,300000,0.00
signup_date,0,300000,0.00
country,0,300000,0.00
city,0,300000,0.00
age_group,0,300000,0.00
customer_type,0,300000,0.00
is_premium,0,300000,0.00
premium_start_date,201086,300000,67.03
acquisition_channel,0,300000,0.00
device_type,0,300000,0.00


`premium_start_date` column is missing **67.03%** of its value which is not an issues.
**it means 67.03% of our users are not Premium users**  

In [18]:
users["user_id"].is_unique

True

`user_id` column has no duplicated id

### **Country Distribution**

In [22]:
users_distribution_by_country = (
    users
        .groupby("country")
        .agg(
            users_count = ("user_id", "count")
        )
        .sort_values("users_count", ascending = False)
)
users_distribution_by_country["percentage"] = (users_distribution_by_country["users_count"] * 100 / users_distribution_by_country["users_count"].sum()).round(2)
users_distribution_by_country.to_csv("../../output/users_distribution_by_country.csv")
users_distribution_by_country

,users_count,percentage
country,,
Egypt,95752,31.92
Saudi Arabia,69090,23.03
UAE,54109,18.04
Kuwait,35958,11.99
Jordan,24171,8.06
Qatar,11902,3.97
Bahrain,9018,3.01


### **City Distribution**

In [ ]:
users_distribution_by_city = (
    users
        .groupby(["country", "city"])
        .agg(
            users_count = ("user_id", "count")
        )
        .sort_values("country")
)
users_distribution_by_city["percentage"] = (users_distribution_by_city["users_count"] / users_distribution_by_city["users_count"].sum()).round(4)
users_distribution_by_city.to_csv("../../output/users_distribution_by_city.csv")
users_distribution_by_city

users_count  percentage
country      city                                
Bahrain      Manama              4475      0.0149
             Muharraq            4543      0.0151
Egypt        Alexandria         13600      0.0453
             Assiut             13816      0.0461
             Cairo              13728      0.0458
             Giza               13710      0.0457
             Mansoura           13621      0.0454
             Tanta              13711      0.0457
             Zagazig            13566      0.0452
Jordan       Amman               8036      0.0268
             Irbid               8104      0.0270
             Zarqa               8031      0.0268
Kuwait       Kuwait City        11994      0.0400
             Farwaniya          11982      0.0399
             Hawalli            11982      0.0399
Qatar        Al Wakrah           5907      0.0197
             Doha                5995      0.0200
Saudi Arabia Dammam             13820      0.0461
             Jeddah             13829      0.0461
             Mecca              14002      0.0467
             Medina             13688      0.0456
             Riyadh             13751      0.0458
UAE          Dubai              13561      0.0452
             Abu Dhabi          13492      0.0450
             Ajman              13496      0.0450
             Sharjah            13560      0.0452

### **Age Group Distribution**

In [ ]:
users_distribution_by_age_group = (
    users
        .groupby("age_group")
        .agg(
            users_count = ("user_id", "count")
        )
        .sort_values("age_group")
)
users_distribution_by_age_group["percentage"] = (users_distribution_by_age_group["users_count"] / users_distribution_by_age_group["users_count"].sum()).round(4)
users_distribution_by_age_group.to_csv("../../output/users_distribution_by_age_group.csv")
users_distribution_by_age_group


,users_count,percentage
age_group,,
18-24,54003,0.1800
25-34,123435,0.4114
35-44,71856,0.2395
45-54,32858,0.1095
55+,17848,0.0595


### **Customer Type Distribution**

In [40]:
users_distribution_by_customer_type = (
    users
        .groupby("customer_type")
        .agg(
            users_count = ("user_id", "count")
        )
)
users_distribution_by_customer_type["percentage"] = (users_distribution_by_customer_type["users_count"] / users_distribution_by_customer_type["users_count"].sum()).round(4)
users_distribution_by_customer_type.to_csv("../../output/users_distribution_by_customer_type.csv")
users_distribution_by_customer_type

,users_count,percentage
customer_type,,
New,183004,0.61
Returning,116996,0.39


### **Premium Status Distribution**

In [43]:
users_distribution_by_premium_status = (
    users
        .groupby("is_premium")
        .agg(
            users_count = ("user_id", "count")
        )
)
users_distribution_by_premium_status["percentage"] = (users_distribution_by_premium_status["users_count"] / users_distribution_by_premium_status["users_count"].sum()).round(4)
users_distribution_by_premium_status.to_csv("../../output/users_distribution_by_premium_status.csv")
users_distribution_by_premium_status

,users_count,percentage
is_premium,,
False,201086,0.6703
True,98914,0.3297


### **Acquisition Channel Distribution**

In [45]:
users_distribution_by_acquisition_channel = (
    users
        .groupby("acquisition_channel")
        .agg(
            users_count = ("user_id", "count")
        )
        .sort_values("users_count", ascending=False)
)
users_distribution_by_acquisition_channel["percentage"] = (users_distribution_by_acquisition_channel["users_count"] / users_distribution_by_acquisition_channel["users_count"].sum()).round(4)
users_distribution_by_acquisition_channel.to_csv("../../output/users_distribution_by_acquisition_channel.csv")
users_distribution_by_acquisition_channel

,users_count,percentage
acquisition_channel,,
Organic Search,83804,0.2793
Direct,66313,0.2210
Social Media,54079,0.1803
Paid Search,45085,0.1503
Referral,26855,0.0895
Email,14994,0.0500
Influencers,8870,0.0296


### **Device Type Distribution**

In [47]:
users_distribution_by_device_type = (
    users
        .groupby("device_type")
        .agg(
            users_count = ("user_id", "count")
        )
)
users_distribution_by_device_type["percentage"] = (users_distribution_by_device_type["users_count"] / users_distribution_by_device_type["users_count"].sum()).round(4)
users_distribution_by_device_type.to_csv("../../output/users_distribution_by_device_type.csv")
users_distribution_by_device_type

,users_count,percentage
device_type,,
Desktop,71753,0.2392
Mobile,204275,0.6809
Tablet,23972,0.0799


### **Monthly Users Signup Trend**

In [ ]:
users_monthly_signup = (
    users
        .assign(
            Year = lambda df: df["signup_date"].dt.year,
            Month_Number = lambda df: df["signup_date"].dt.month,
            Month_Name = lambda df: df["signup_date"].dt.month_name(),
            Month_Name_Short = lambda df: df["signup_date"].dt.strftime("%b"),
            Date_Label = lambda df: df["signup_date"].dt.strftime("%b-%y")
        )
        .groupby(["Year", "Month_Number", "Month_Name", "Month_Name_Short", "Date_Label"])
        .agg(
            users_count = ("user_id", "count")
        )
)
users_monthly_signup["percentage"] = (users_monthly_signup["users_count"] / users_monthly_signup["users_count"].sum()).round(4)
users_monthly_signup.to_csv("../../output/users_monthly_signup.csv")
users_monthly_signup

users_count  \
Year Month_Number Month_Name Month_Name_Short Date_Label                
2025 1            January    Jan              Jan-25            25634   
     2            February   Feb              Feb-25            23067   
     3            March      Mar              Mar-25            25509   
     4            April      Apr              Apr-25            24991   
     5            May        May              May-25            25737   
     6            June       Jun              Jun-25            24708   
     7            July       Jul              Jul-25            25372   
     8            August     Aug              Aug-25            25483   
     9            September  Sep              Sep-25            24571   
     10           October    Oct              Oct-25            25572   
     11           November   Nov              Nov-25            24673   
     12           December   Dec              Dec-25            24683   

                                                          percentage  
Year Month_Number Month_Name Month_Name_Short Date_Label              
2025 1            January    Jan              Jan-25          0.0854  
     2            February   Feb              Feb-25          0.0769  
     3            March      Mar              Mar-25          0.0850  
     4            April      Apr              Apr-25          0.0833  
     5            May        May              May-25          0.0858  
     6            June       Jun              Jun-25          0.0824  
     7            July       Jul              Jul-25          0.0846  
     8            August     Aug              Aug-25          0.0849  
     9            September  Sep              Sep-25          0.0819  
     10           October    Oct              Oct-25          0.0852  
     11           November   Nov              Nov-25          0.0822  
     12           December   Dec              Dec-25          0.0823